sycl/CMakeLists.txt

```diff
@@ -169,37 +168,49 @@ include(AddBoostMp11Headers)
 file(GLOB_RECURSE HEADERS_IN_SYCL_DIR CONFIGURE_DEPENDS "${sycl_inc_dir}/sycl/*")
 file(GLOB_RECURSE HEADERS_IN_CL_DIR CONFIGURE_DEPENDS "${sycl_inc_dir}/CL/*")
 file(GLOB_RECURSE HEADERS_IN_STD_DIR CONFIGURE_DEPENDS "${sycl_inc_dir}/std/*")
-string(REPLACE "${sycl_inc_dir}" "${SYCL_INCLUDE_BUILD_DIR}"
-  OUT_HEADERS_IN_SYCL_DIR "${HEADERS_IN_SYCL_DIR}")
-string(REPLACE "${sycl_inc_dir}/CL" "${SYCL_INCLUDE_BUILD_DIR}/sycl/CL"
-  OUT_HEADERS_IN_CL_DIR "${HEADERS_IN_CL_DIR}")
-string(REPLACE "${sycl_inc_dir}" "${SYCL_INCLUDE_BUILD_DIR}"
-  OUT_HEADERS_IN_STD_DIR "${HEADERS_IN_STD_DIR}")
-
-# Copy SYCL headers from sources to build directory
+# configure_file templates shouldn't be installed
+# We overglob and then remove our configure_file templates, because
+# file(GLOB_RECURSE doesn't have a usable EXCLUDE REGEX flag
+list(FILTER HEADERS_IN_SYCL_DIR EXCLUDE REGEX "\.in$")
+list(FILTER HEADERS_IN_CL_DIR EXCLUDE REGEX "\.in$")
+list(FILTER HEADERS_IN_STD_DIR EXCLUDE REGEX "\.in$")
+
+# copy the headers to the build directory...
+set(SYCL_OUT_HEADERS)
+foreach(F IN LISTS HEADERS_IN_SYCL_DIR HEADERS_IN_CL_DIR HEADERS_IN_STD_DIR)
```

Comment:
Could you please clarify if there are any problems with having `.in` templates in the build directory? 
Can we instead copy all headers like we are doing without the patch, but update `install` target to exclude `.in` files:
```
 install(DIRECTORY "${sycl_inc_dir}/sycl" DESTINATION ${SYCL_INCLUDE_DIR} COMPONENT sycl-headers PATTERN "*.in" EXCLUDE)
```

Reply:
We can - but it's a conceptual issue. I don't see the advantage of filtering late (when we install) vs filtering early and not copying the template source to the build directory.

I think the clean solution would actually be to avoid globbing altogether, and simply maintain a static list of installed headers and a separate list of templated header. When a header is added to the source tree, it would also need a corresponding change. This manual updating avoids lots of configuration time vs build time update ordering bugs, and makes people think about what they're doing. That's a later change I'd like to make, but in the present patch I want to reinforce that source templates belong in the source tree and that the generated headers should end up in the build tree.

I wonder if we need to copy the headers to the build tree *at all*, but I have a few more build system cleanups I'll be working on and can discuss them when I do so.

Reply:
As an alternative we can move config/`.in` files in a dedicated directory which won't be copied.
I'm not an expert here, but having a cmake target for each header file sounds suboptimal and may lead to issues in incremental build(aka a new header added, but cmake hasn't "reGLOBed" files).